druid-shell/src/platform/windows/keyboard.rs

```diff
@@ -0,0 +1,781 @@
+        _ if code_unit >= 0x20 => {
+            if let Some(c) = std::char::from_u32(code_unit) {
+                Key::Character(c.to_string())
+            } else {
+                // UTF-16 error, very unlikely
+                Key::Unidentified
+            }
+        }
+        _ => Key::Unidentified,
+    }
+}
+
+/// Get location from virtual key code.
+///
+/// This logic is based on NativeKey::GetKeyLocation from Mozilla.
+fn vk_to_location(vk: VkCode, is_extended: bool) -> Location {
+    match vk as INT {
+        VK_LSHIFT | VK_LCONTROL | VK_LMENU | VK_LWIN => Location::Left,
+        VK_RSHIFT | VK_RCONTROL | VK_RMENU | VK_RWIN => Location::Right,
+        VK_RETURN if is_extended => Location::Numpad,
+        VK_INSERT | VK_DELETE | VK_END | VK_DOWN | VK_NEXT | VK_LEFT | VK_CLEAR | VK_RIGHT
+        | VK_HOME | VK_UP | VK_PRIOR => {
+            if is_extended {
+                Location::Standard
+            } else {
+                Location::Numpad
+            }
+        }
+        VK_NUMPAD0 | VK_NUMPAD1 | VK_NUMPAD2 | VK_NUMPAD3 | VK_NUMPAD4 | VK_NUMPAD5
+        | VK_NUMPAD6 | VK_NUMPAD7 | VK_NUMPAD8 | VK_NUMPAD9 | VK_DECIMAL | VK_DIVIDE
+        | VK_MULTIPLY | VK_SUBTRACT | VK_ADD | VK_ABNT_C2 => Location::Numpad,
+        _ => Location::Standard,
+    }
+}
+
+impl KeyboardState {
+    /// Create a new keyboard state.
+    ///
+    /// There should be one of these per window. It loads the current keyboard
+    /// layout and retains some mapping information from it.
+    pub fn new() -> KeyboardState {
+        unsafe {
+            let hkl = GetKeyboardLayout(0);
+            let key_vals = HashMap::new();
+            let dead_keys = HashSet::new();
+            let stash_vk = None;
+            let stash_utf16 = Vec::new();
+            let has_altgr = false;
+            let mut result = KeyboardState {
+                hkl,
+                key_vals,
+                dead_keys,
+                has_altgr,
+                stash_vk,
+                stash_utf16,
+            };
+            result.load_keyboard_layout();
+            result
+        }
+    }
+
+    /// Process one message from the platform.
+    ///
+    /// This is the main interface point for generating cooked keyboard events
+    /// from raw platform messages. It should be called for each relevant message,
+    /// which comprises: `WM_KEYDOWN`, `WM_KEYUP`, `WM_CHAR`, `WM_SYSKEYDOWN`,
+    /// `WM_SYSKEYUP`, `WM_SYSCHAR`, and `WM_INPUTLANGCHANGE`.
+    ///
+    /// As a general theory, many keyboard events generate a sequence of platform
+    /// messages. In these cases, we stash information from all messages but the
+    /// last, and generate the event from the last (using `PeekMessage` to detect
+    /// that case). Mozilla handling is slightly different; it actually tries to
+    /// do the processing on the first message, fetching the subsequent messages
+    /// from the queue. We believe our handling is simpler and more robust.
```

Comment:
I think it's great to write stuff down here. Something that would be particularly helpful is an example of a keyboard event that generates multiple messages.

Reply:
I added a couple examples.